okapi-core/src/main/java/org/folio/okapi/service/impl/DockerModuleHandle.java

```diff
@@ -373,64 +348,40 @@ private int getExposedPort(JsonObject b) {
     return exposedPort;
   }
 
-  private void prepareContainer(Handler<AsyncResult<Void>> startFuture) {
-    getImage(res1 -> {
-      if (res1.failed()) {
-        logger.warn("getImage failed 1 : {}", res1.cause().getMessage());
-        startFuture.handle(Future.failedFuture(res1.cause()));
-        return;
-      }
-      if (hostPort == 0) {
-        startFuture.handle(Future.failedFuture(messages.getMessage("11300")));
-        return;
-      }
+  private Future<Void> prepareContainer() {
+    if (hostPort == 0) {
+      return Future.failedFuture(messages.getMessage("11300"));
+    }
+    return getImage().compose(res1 -> {
       int exposedPort;
       try {
-        exposedPort = getExposedPort(res1.result());
-      } catch (Exception ex) {
-        startFuture.handle(Future.failedFuture(ex));
-        return;
+        exposedPort = getExposedPort(res1);
+      } catch (Exception e) {
+        logger.warn("{}", e.getMessage(), e);
+        return Future.failedFuture(e);
       }
-      createContainer(exposedPort, res2 -> {
-        if (res2.failed()) {
-          startFuture.handle(res2);
-          return;
-        }
-        startContainer(res3 -> {
-          if (res3.failed()) {
-            deleteContainer(x -> startFuture.handle(res3));
-            return;
-          }
-          getContainerLog(res4 -> {
-            if (res4.failed()) {
-              this.stop(x -> startFuture.handle(res4));
-              return;
-            }
-            startFuture.handle(res4);
-          });
-        });
-      });
+      return createContainer(exposedPort)
+          .compose(res2 -> startContainer()
+              .onFailure(cause -> deleteContainer())
+              .compose(res3 -> getContainerLog())
+      );
     });
   }
 
   @Override
-  public void start(Handler<AsyncResult<Void>> startFuture) {
+  public Future<Void> start() {
+    Future<Void> future = Future.succeededFuture();
     if (dockerPull) {
-      pullImage(res -> prepareContainer(startFuture));
-    } else {
-      prepareContainer(startFuture);
+      future = future.compose(x -> pullImage());
     }
+    return future.compose(x -> prepareContainer());
```

Comment:
if (dockerPull) {
      return pullImage().compose(x -> prepareContainer());
    }
    return prepareContainer();

Reply:
OK. The reason for existing layout was to mention prepareContainer only once.